File: python/39_combination-sum.py
# Recursion
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        # Define function
        def find_combinations(results, combination, candidates, target, start_index):
            if target == 0:
                results.append(combination[:])
                return

            for i in range(start_index, len(candidates)):
                if candidates[i] > target:
                    break

                combination.append(candidates[i])
                find_combinations(results, combination, candidates, target - candidates[i], i)
                combination.pop()

        results = []
        combination = []

        # Check inputs
        if not candidates or len(candidates) == 0:
            return results

        # Sort candidates
        candidates = sorted(candidates)

        find_combinations(results, combination, candidates, target, 0)
        return results


#---------------------------------------------------------------------------------------#


# Dynamic Programming
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        dp = [[] for _ in range(target + 1)]
        dp[0] = [[]]

        for c in sorted(candidates):
            for i in range(target + 1):
                if i >= c:
                    for d in dp[i - c]:
                        dp[i].append(d + [c])

        return dp[-1]

File: go/39_combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
	dp := make([][][]int, target+1)
	dp[0] = [][]int{{}}

	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i] < candidates[j]
	})

	for _, c := range candidates {
		for i := 0; i < target+1; i++ {
			if i >= c {
				for _, d := range dp[i-c] {
					tmp := make([]int, len(d))
					copy(tmp, d)
					tmp = append(tmp, c)
					dp[i] = append(dp[i], tmp)
				}
			}
		}
	}
	return dp[len(dp)-1]
}